### ISSUE TITLE
Seq<A>.Take(n) returns sequence of null values if n > Count

### ISSUE THREAD

**Issue by huntwj** [louthy/language-ext#795]
```cs
var s = Seq<MyType>();
var firstFive = s.Take(5);

Assert.Equal(s, firstFive);

// Expected:
// firstFive is []
// Observed:
// firstFive is [null, null, null, null, null]
```

I'm used to Take meaning Take "up to n if they exist." Is this the correct behavior?

**Comment by bender2k14** [comment]
Your test passes for me using [the commit with tag `3.4.11`](https://github.com/louthy/language-ext/tree/3.4.11).

What version are you using?

**Comment by huntwj** [comment]
I am using 3.4.15 released 5/14/2020.

Perhaps this is a regression in later versions?

**Comment by huntwj** [comment]
I actually do not see that version in the GitHub tag list. Here is what VS is showing me:

![image](https://user-images.githubusercontent.com/2975468/84949964-f16f1180-b0a2-11ea-802e-ee9eb395ace0.png)

**Comment by louthy** [comment]
Certainly a big merge happened to bring in the traverse work, so perhaps something fell through the cracks.  If you can confirm you see this with the latest `beta` then I'll do a bit of digging tomorrow.

**Comment by bender2k14** [comment]
@huntwj, I am unable to reproduce the behavior you described.  All the commits I tested behaved as expected.

In particular, here is my execution of your code when depending on version 3.4.15 of the NuGet package.
![2020-06-17_20-00-02_129_devenv](https://user-images.githubusercontent.com/34664007/84965768-6bb88980-b0d5-11ea-912b-247bd5a0ff94.png)

**Comment by huntwj** [comment]
I'll have to go back in my commits to see if I can find something that still has the problem. I changed the area of "problem code" to use `Lst<MyType>` instead of `Seq<MyType>` and things seemed to work better. I'm up against a deadline here, so I don't have too much extra time right now, but I'll try to track it down once we finish with this release.

Thank you all for your efforts in tracking this down, even if it does turn out to be a phantom. :)

**Comment by huntwj** [comment]
Ok, I needed a break, so I decided to throw together a sample. Here is the demo repository:

https://github.com/huntwj/seq-take-bug.git

And the more "complex" code required to replicate:

```cs
Seq<MyType> emptyAfterLinq = from item in LongSeq() // LongSeq returns several items
                             where item.Value > 5   // but all are filtered out by where clause
                             select item;
Console.WriteLine(emptyAfterLinq);                  // Outputs: []

Seq<MyType> take5 = emptyAfterLinq.Take(5); 
Console.WriteLine(take5);                           // Outputs: [, , , , ]
```
The problem doesn't show up until after I do the Linq query. As @bender2k14 demonstrated above, just creating an empty `Seq` wasn't sufficient. Doing a Ling query that results in an empty `Seq`, however, does the trick.

**Comment by bender2k14** [comment]
Ah, very good @huntwj.

Here is a MWE of that.  The bug is in `3.4.24-beta` c1b94568b8c75c2da4c00dfee9ba70e6214f8db0.

```
[Fact]
public void CompletelyFilteredGeneratedSequence()
{
    var actual =
        generate(1, _ => unit)
        .Filter(_ => false);
    Assert.Equal(Seq<Unit>(), actual);  // Passes
}

[Fact]
public void TakeOneFromCompletelyFilteredGeneratedSequence()
{
    var actual =
        generate(1, _ => unit)
        .Filter(_ => false)
        .Take(1);
    Assert.Equal(Seq<Unit>(), actual);  // Fails
}
```

**Comment by bender2k14** [comment]
The bug is in here.  I don't understand what all the cases are for in the `else` branch.
https://github.com/louthy/language-ext/blob/3fbe724a2c1ce0cd3fb1d1c1a29fc279d395bf31/LanguageExt.Core/DataTypes/Seq/SeqLazy.cs#L368-L399

**Comment by louthy** [comment]
I'll take a look

**Comment by louthy** [comment]
Fixed in `3.4.24-beta`
